File: 5/py/supply_stacks.py
import copy

from stacks import stacks

def run():
    file_data = read_file('supply_stacks.txt')

    print(sort_stacks(file_data))
    print(rearrangement(file_data))

def read_file(path):
    with open(path) as file:
        return file.read().strip().split('\n')

def make_output(stack):
    output = ''

    for item in stack:
        output += item[0]

    return output

def sort_stacks(sorting_data):
    stack = copy.deepcopy(stacks)

    for line in sorting_data:
        instructions = line.split(' ')
        quantity, origin, destination = int(instructions[1]), int(instructions[3]) - 1, int(instructions[5]) - 1

        for _ in range(quantity):
            stack[destination].insert(0, stack[origin].pop(0))
            

    return make_output(stack)

def rearrangement(sorting_data):
    stack = copy.deepcopy(stacks)

    for line in sorting_data:
        instructions = line.split(' ')
        quantity, origin, destination = int(instructions[1]), int(instructions[3]) - 1, int(instructions[5]) - 1

        for item in reversed(range(quantity)):
            stack[destination].insert(0, stack[origin].pop(item))
            
    return make_output(stack)
    

File: 5/rb/supply_stacks.rb
require './stacks.rb'

def run
    file_data = File.read('supply_stacks.txt').split("\n")

    sort_stacks(file_data)
end

def sort_stacks(sorting_data)
    # doesn't work
    stack = stacks()

    sorting_data.each do |line|
        instructions = line.split("\s")
        quantity, from, to = instructions[1].to_i, instructions[3].to_i - 1, instructions[5].to_i - 1
        
        for _ in 0..quantity do
            if stack[from] != ''
                stack[from], popped = stack[from][1..-1], stack[from][0]
                stack[to] = popped + stack[to]
            end
        end
    end

    stack.each do |row|
        puts row[0]
    end
end